lib/auth/auth_with_roles.go

```diff
@@ -5300,13 +5299,29 @@ func (a *ServerWithRoles) GetAppSession(ctx context.Context, req types.GetAppSes
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	// Users can only fetch their own app sessions.
-	if err := a.currentUserAction(session.GetUser()); err != nil {
-		if err := a.action(apidefaults.Namespace, types.KindWebSession, types.VerbRead); err != nil {
-			return nil, trace.Wrap(err)
+
+	authErr := a.action(apidefaults.Namespace, types.KindWebSession, types.VerbRead)
+	if authErr == nil {
+		return session, nil
+	}
+
+	// Users can fetch their own app sessions without secrets.
+	if err := a.currentUserAction(session.GetUser()); err == nil {
+		// TODO (Joerger): DELETE IN 17.0.0
+		// App Session secrets should not be returned to the user. We only do this
+		// here for backwards compatibility with `tsh proxy azure`, which uses the
+		// app session key to sign JWT tokens with Azure claims. This check means
+		// that `tsh proxy azure` will fail for old clients when used with Per-session
+		// MFA or Hardware Key support, which is planned for release in v16.0.0.
+		identity := a.context.Identity.GetIdentity()
+		if identity.MFAVerified == "" || identity.PrivateKeyPolicy != keys.PrivateKeyPolicyWebSession {
+			return session, nil
 		}
```

Comment:
Is there any practical way a user may "downgrade" their own session, removing the `MFAVerified` bit, yet gaining access to the private keys here?

Reply:
This shouldn't be possible, no. There is no way for a user to update their own session and replace its certs. Even if they could, the session would not be MFA verified so there would be no harm done.

Or is your concern that we _should_ let users downgrade sessions? If the user is using a new client, there would be no reason for them to access the private key. In v17.0.0 we will always return the session without secrets.

Reply:
This is probably a theoretical case, but the oddness here is that we grant more privileges (access to the private key) to a strictly weaker identity (without MFA verified bit). Hence I wonder if the user can force that downgrade, effectively increasing their access. I don't know how would that happen, hence my question... but I figure the answer is "no".

Reply:
Gotcha. The way to think about this is that the stricter identities grant more privilege, hence they need to be guarded more strictly. Even if the user could downgrade the privilege, it would no longer need to be guarded as strictly.